### DISCUSSION TITLE
Migrate from 2.0 to 3.4

### DISCUSSION THREAD

**Discussion by jessecureton** [papermerge/papermerge-core#589] [General]
I've been trying to get my papermerge 2.0.1 installation updated to the new 2.1 release, and having lots of issues. The root of the issue seems to be that the `core_user.id` column in my old database is an integer, when some migrations are expecting a `uuid` field. This means that 9 migrations fail to apply when first launching, and from there everything just seems to cascade and I can't sign in due to more errors with the core_user table when attempting a login.

Is this an upgrade that's expected to work? Right now things seem pretty borked, but I'm happy to pull a DB backup and help gather more data if this is something that's expected to work

**Comment by ciur** [comment]
Thank you for opening this ticket!

In short, migration from 2.0 to 2.1 is more complex then just DB changes. Just DB part will not work even if you succeed in fixing db migrations; why? because 2.1 introduces new DB model for document version which wasn't there before :).

A picture will help understanding why DB changes won't work:

![schema-changes](https://user-images.githubusercontent.com/24827601/210509157-e6deb4bb-d208-4bcb-92fc-fc39167cb38f.png)

As you see from the picture, the "Document Version" model was introduced in 2.1 (before that, document versions were "stored" only on file-system). There is literally no information about document version(s) in DB before 2.1 version.

Another important change is actually what information is stored on the file-system level i.e. what and how OCRed data is stored
on file system. "OCR data" is much more then just plain text characters - extracted "OCR data" also stores coordinates of each extracted character (this info has a fancy name - [hocr](https://en.wikipedia.org/wiki/HOCR) format).

Long story short - at this moment you don't need to do anything.
Please give me couple of weeks and I will document and provide necessary tools on how to migrate documents from 2.0 to 2.1.

The only caveat is that your metadata information will not be transferred - as 2.1 currently does has the concept of matadata. Metadata will be re-introduced in later releases.

Again, I will provide soon necessary tools and documentation on how to migrate from 2.0 to 2.1.

**Comment by jessecureton** [comment]
Hi @ciur! Thanks again for all your work on this. I saw the newest release and wanted to follow up on this and see if there has been any progress on a migration path from 2.0->2.1?

**Comment by duracell** [comment]
Hey @ciur, any news about the migration to 2.1? I'm currently using the old linuxserver image (2.0x), but would love to switch over and upgrade to a new version. Should it work?

**Comment by ReneHezser** [comment]
Hi @ciur 
I am looking for upgrade instructions from 2.0 to the latest 3.2 version and was wondering if there is an updated upgrade instruction available. The current installation is running in a container and is installed with Docker Compose.

**Comment by duracell** [comment]
It really seems like a mess :'/ Breaking update from 2 → 2.1 and another breaking update from 2.1 → 3.0…
No updates for old versions, no migration to new version and totally mixed features among different versions, without any explanation if these features will be return or not. The great sharing feature was the reason why I choose papermerge over Paperless-ng(x), but it seems the new version just getting the features of paperless and drops the unique features. :(
I really grateful, and thanking ciur for the fantastic work of this software, but this great work makes me even more sad that I'm using an old version and can't update. =/

**Comment by ciur** [comment]
> It really seems like a mess :'/ Breaking update from 2 → 2.1 and another breaking update from 2.1 → 3.0…

Yes, I agree, it is confusing and painful to have those breaking incompatibilities. When I started this project I had no idea how to write a DMS software and as result I made terrible mistakes in design of the application. The outcome was very painful maintenance of the code base. Also adding new features was hard because of the base structure on which the entire project was built.

Because I have big plans for the project and because the project is still relatively new I've decided to break backward compatibility for the sake of fixing design errors in early stage of the project.

The good part is that since 3.0 there won't be any breaking changes in terms of how files are stored on the disk and database schema: this means that once in 3.0 you will be able to update to 3.1, 3.2, 3.3,... without any problem.

> No updates for old versions, no migration to new version and totally mixed features among different versions, without any explanation if these features will be return or not.

All great features present in 2.0 (like sharing files between users, metadata, automates) will be back. Both "sharing of the files between users" and "metadata" will be back this year (2024).

> but it seems the new version just getting the features of paperless and drops the unique features

In short term yes - because of large code base refactoring inevitably some of the non essential features need to be "temporary removed". But in long term, features from 2.0 will be back in significantly better shape. 

> I'm using an old version and can't update.

I am currently in deep dive mode with 3.3; with 3.3. released, I will dedicate time for documenting migration from 2.0 and 2.1 directly into 3.x.

**Comment by ReneHezser** [comment]
Thank you very much for you effort on this great project.

> I am currently in deep dive mode with 3.3; with 3.3. released, I will dedicate time for documenting migration from 2.0 and 2.1 directly into 3.x.

That is the important information. At least I can and will wait some more time to upgrade later. Thanks.

**Comment by duracell** [comment]
Thanks for the fast and informative reply and the promise of a stable future. I think that is really important for a DMS.

> I am currently in deep dive mode with 3.3; with 3.3. released, I will dedicate time for documenting migration from 2.0 and 2.1 directly into 3.x.

That sounds great, and then I will wait instead of trying to migrate everything to another software, as there is currently no real pressure except the missing updates.

**Comment by duracell** [comment]
@ciur any news about upgrade/migrate an installation to a new version?

**Comment by duracell** [comment]
@ciur So, is the unpinning a message?
I just would like to know if it's useful to wait or if I just should manual migrate to new version/software.
A short info would be helpful, even if you just say you don't have the resources to do this documentation.

**Comment by ciur** [comment]
Migrate to the newest version  which is 3.3.
I don't have resources to do this documentation.

**Comment by ciur** [comment]
@duracell Are you on 2.0?

Is the migration problem still relevant ?

The plan now is to provide migration scripts from 2.0 to 3.4. So I will update this ticket title as "migrate from 2.0 to 3.4".
Will be ok for you to migrate directly 3.4 ?

**Comment by duracell** [comment]
Hey, my Papermerge is on version 2.0, but unfortunately, it’s no longer relevant for me. After the comment I spent several hours migrating to a different solution and completely rebuild my scanning workflow.  I see the complexity and lower priority of the whole migration topic, so I didn’t want to wait hoping or annoy you with asking again.
But thanks for the offer :)

**Comment by ciur** [comment]
Is there anybody interested in migrating from 2.0 to 3.4 ? If yes, now is the moment to raise your hand.
If there will be no interest until June 2025, I will close the topic.

**Reply by ReneHezser** [reply]
I am. Still running 2.0 and would like to upgrade.

**Reply by ciur** [reply]
Good to know. I will provide migration instructions (+ required apps) by the end of these month (February 2025).

**Reply by ciur** [reply]
@ReneHezser What database do you use ? Which version ?

**Reply by ReneHezser** [reply]
@ciur Thanks for your work. I am using the linuxserver image [linuxserver/papermerge:v2.0.1-ls60](https://github.com/linuxserver-archive/docker-papermerge/tree/v2.0.1-ls60) 
My papermerge.config contains and the mapped data folder contains a papermerge.db file
```
DBHOST = "mariadb"
DBNAME = "papermerge"
```
The package **libmariadb3** is installed in that image.

**Reply by ciur** [reply]
@ReneHezser could you pls provide me with your configuration? i.e. docker compose file + papermerge.conf.py file so that i can sort of reproduce your setup locally?

**Reply by ReneHezser** [reply]
Hey @ciur 
happy to help and provide everything needed.

My compose file (parts of it).
```docker
# version: '3.8'
services:
  papermerge:
    image: linuxserver/papermerge:v2.0.1-ls60
    container_name: papermerge
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Berlin
      - DOCKER_MODS=linuxserver/mods:papermerge-multilangocr
      - OCRLANG=deu
    volumes:
      - /docker/papermerge/config:/config
      - /docker/papermerge/data:/data
    ports:
      - 8090:8000
    restart: unless-stopped
    labels: 
      - com.centurylinklabs.watchtower.enable=true
```

papermerge.conf.py
```python
#DBTYPE = "mariadb"  # Uncomment this to enable an external DB instead of local SQLite, refer to Papermerge docs
DBUSER = "root"
DBPASS = "root"
DBHOST = "mariadb"
DBNAME = "papermerge"

MEDIA_DIR = "/data/media"
STATIC_DIR = "/app/papermerge/static"
MEDIA_URL = "/media/"
STATIC_URL = "/static/"

OCR_DEFAULT_LANGUAGE = "deu"

OCR_LANGUAGES = {
    "deu": "Deutsch",
    "eng": "English",
}

LANGUAGE_CODE = "de-DE"
BINARY_STAPLER = "/usr/local/bin/stapler"
```

settings.py
```python
from .base import *  # noqa

DEBUG = False
# debug variable in templates is available only if INTERNAL_IPS are set
# to a not empty list
INTERNAL_IPS = [
    '127.0.0.1',
]

ALLOWED_HOSTS = ['*']

INSTALLED_APPS.extend(
    ['mod_wsgi.server', ]
)

# DATABASES = {
#     'default': {
#         'ENGINE': 'django.db.backends.postgresql',
#         'NAME': 'dbname',
#         'USER': 'dbuser',
#         'PASSWORD': 'dbpass',
#         'HOST': 'db',
#         'PORT': 5432,
#     },
# }

# DATABASES = {
#     'default': {
#         'ENGINE': 'django.db.backends.mysql',
#         'NAME': 'dbname',
#         'USER': 'dbuser',
#         'PASSWORD': 'dbpass',
#         'HOST': 'db',
#         'PORT': 3306,
#     },
# }

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.sqlite3",
        "NAME": "/data/papermerge.db",
    }
}

LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'file_worker': {
            'class': 'logging.FileHandler',
            'filename': 'worker.log',
        },
        'file_app': {
            'class': 'logging.FileHandler',
            'filename': 'app.log',
        },
    },
    'loggers': {
        'mglib': {
            'handlers': ['file_app'],
            'level': 'DEBUG'
        },
        'papermerge': {
            'handlers': ['file_app'],
            'level': 'DEBUG'
        },
        'celery': {
            'handlers': ['file_worker'],
            'level': 'INFO'
        },
    },
}

CELERY_BROKER_URL = "redis://localhost"
CELERY_BROKER_TRANSPORT_OPTIONS = {}
CELERY_RESULT_BACKEND = "redis://localhost"
```

Thanks a lot.

**Comment by ciur** [comment]
I was able to start the app. I am using mariadb 11.2.
I think I need a week or two to write the migration utility. Migration utility will be a binary which you will need to run on your platform. 
Couple of more question:
What maria db version are you using?
What platform are you running docker compose on? Is it linux amd64?
Second question am asking because I need to know that  platform to compile the binary for. In worst case scenario you will run migration utility from inside docker container. I will provide instructions. Of course migration utility will be open source, so you will have the option to adjust it to your needs and eventually to compile it yourself for your OS.

Here is the compose file with which I started your setup:
```
services:
  papermerge:
    image: linuxserver/papermerge:v2.0.1-ls60
    container_name: papermerge
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Berlin
      - DOCKER_MODS=linuxserver/mods:papermerge-multilangocr
      - OCRLANG=deu
    volumes:
      - ${PWD}/config:/config
      - ${PWD}/data:/data
    ports:
      - 8090:8000
    restart: unless-stopped
    labels:
      - com.centurylinklabs.watchtower.enable=true
  db:
    image: mariadb:11.2
    restart: always
    environment:
      MARIADB_ROOT_PASSWORD: root
      MARIADB_USER: root
      MARIADB_DATABASE: papermerge
      MARIADB_PASSWORD: root
    ports:
      - 3306:3306
```

**Reply by ReneHezser** [reply]
> What maria db version are you using?

I am not running maria db outside of that container. The comment `#DBTYPE = "mariadb"  # Uncomment this to enable an external DB instead of local SQLite, refer to Papermerge docs` tells me, that I am using local SQLite only.

> What platform are you running docker compose on? Is it linux amd64?

Yes, I am running on `Ubuntu 20.04.6 LTS (GNU/Linux 5.4.0-176-generic x86_64)`. Running inside the container would be fine for me.

> Here is the compose file with which I started your setup:

I don't think you need maria db, as at least I am not using it.

Thanks for your effort!

**Comment by ciur** [comment]
@ReneHezser 

Could you give it a try with  [pmdump v0.1](https://github.com/papermerge/pmdump/releases/tag/v0.1) ?

At this moment instructions are only in README.md file, here: https://github.com/papermerge/pmdump
In short, you need to create a source.yaml file which will instruct pmdump where to copy data from.
Example of source.yaml:

```
app_version: 2.0
media_root: /home/eugen/DockerCompose/linux-server-pmg-2/data/media
database_url: sqlite:///home/eugen/DockerCompose/linux-server-pmg-2/data/papermerge.db
```

Then you need to:
```
pmdump -c source.yaml -f pmg.tar.gz export
```

Which will create pmg.tar.gz file with your data.

Then create a target.yaml:
```
media_root: /home/eugen/DockerCompose/pm3.4-pg/media/
database_url: postgresql://coco:kesha@127.0.0.1:5432/pmg34?sslmode=disable
app_version: 3.4
```

And finally run command (pmg.tar.gz file is the one created by export command):
```
pmdump -c target.yaml -f pmg.tar.gz import
```

Note that only users, documents and folders will be exported. 


For testing source (i.e. Papermerge DMS 2.0) I've used your docker compose. For testing import into the target,
I've tested with following docker compose:
```
services:
  webapp:
    image: papermerge/papermerge:3.4
    environment:
      PAPERMERGE__SECURITY__SECRET_KEY: 12345
      PAPERMERGE__AUTH__USERNAME: admin
      PAPERMERGE__AUTH__PASSWORD: admin
      PAPERMERGE__DATABASE__URL: postgresql://coco:kesha@db:5432/pmg34a4
    ports:
     - "12034:80"
    volumes:
      - ${PWD}/media34a4:/core_app/media
    depends_on:
      - db
  db:
    image: postgres:16.1
    restart: always
    environment:
      - POSTGRES_PASSWORD=kesha
      - POSTGRES_DB=pmg34a4
      - POSTGRES_USER=coco
    volumes:
      - pg_data34a4:/var/lib/postgresql/data   # Store database data
    ports:
      - 5432:5432

volumes:
  pg_data34a4:
```

Notice that local media folder (which should be accessible to pmdump) is mounted unto `/core_app/media` in `papermerge/papermerge:3.4` container.

**Reply by ReneHezser** [reply]
I can try this on Thursday earliest. Will get back with results. Thanks for the guidance and support!

**Reply by ReneHezser** [reply]
A short feedback. I am seeing some files that are missing, after the migration. The pdf and jpg files are in the UI but can't be opened. I need to test some more.
Running 3.4 with SQLite instead of Postgress should not be an issue, I guess.

**Reply by ciur** [reply]
You have PDF and jpg files only?
Do you have other types of files ? (png, tiff?)
Does the export/import work for PDF files?

For jpg files there may be small issue. Indeed they may not open correctly. 
How many jpg files do you have ?

**Reply by ReneHezser** [reply]
I have only 6 jpg and 423 pdfs. No other file formats.
The files are missing after importing. There is no file for the name in the docvers structure

**Reply by ciur** [reply]
I guess 6 jpeg documents you can transfer, between instances, manually.

**Reply by ReneHezser** [reply]
But there are numerous pdfs, that are also not available.

**Reply by ciur** [reply]
Do I understand correctly that files physically are not in media root on the target?
What I mean is following. Let's say PDF y1 does not show correctly in 3.4 after import. But you still can see the
UUID of the document right ? (it is the one visible in URL). Let's say you can see the UUID of the document y1 - uuid_y1.
Please check (in 3.4) its document version: you can do that in SQL with:

```
SELECT id FROM document_versions WHERE document_id = <uuid_y1>
```
Note that uuid_y1 may or may not contain dashes (depends on database type).
Once you know DOCUMENT VERSION UUID you can check if corresponding file is present in media root:

```
<media_root>/ docvers / VERSION_UUID[0:2] / VERSION_UUID[2:4] / VERSION_UUID /  
```

Also check if document version with same UUID is present in tar.gz.
Does `pmdump` issue some warnings/messages during export/import ?

PS: The UUID that you see in URL is Document ID. Don't confuse it with DocumentVersion ID.  Each document must have at least one associated (document) version ID.

**Reply by ReneHezser** [reply]
- There is no entry in the database for the document_id `ee3664a3-d1bc-482d-88bf-ef7190d334da`.
- I see two files for the filename in the folders: `./docvers/cc/6a/cc6a5a4b-c8f0-4da5-a86e-43525ac920fc/<filename>.pdf` and `./docvers/10/b9/10b97c1d-0f29-4d30-bf00-ad8c76ee1572/<filename>.pdf`
- The version id is `cc6a5a4b-c8f0-4da5-a86e-43525ac920fc`, according to the UI
- The export.yml has this data about the document
```yaml
<filename>.pdf:
    id: ee3664a3-d1bc-482d-88bf-ef7190d334da
    legacy_id: 160
    title: <filename>.pdf
    node_type: document
    versions:
        - id: 10b97c1d-0f29-4d30-bf00-ad8c76ee1572
          legacy_id: 0
          number: 0
          file_name: <filename>.pdf
          pages: []
        - id: cc6a5a4b-c8f0-4da5-a86e-43525ac920fc
          legacy_id: 0
          number: 1
          file_name: <filename>.pdf
          pages: []
    file_name: <filename>.pdf
    page_count: 1
```

There were no warnings or errors.

**Reply by ReneHezser** [reply]
I've now copied all missing documents manually. Thank you for your support.
